instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/AlternativeRouteSelectionTest.kt

```diff
@@ -42,124 +33,105 @@ class AlternativeRouteSelectionTest : BaseTest<BasicNavigationViewActivity>(
-            }
-            mockLocationReplayerRule.playRoute(route)
-
-            mapboxNavigation.setRoutes(directionsResponse.routes())
-        }
     }
 
     @After
     fun tearDown() {
-        initIdlingResource.unregister()
+        mapboxNavigation.onDestroy()
     }
 
     @Test
-    fun selectAlternateRoute() {
-        addRouteLine()
-        routeProgressIdlingResource.register()
+    fun expect_route_apis_to_be_updated_with_alternatives() {
+        // Initialize with a set of routes with alternatives. Later we will
+        // update the routes with a new primary route and verify it is all set.
+        setupRouteWithAlternatives()
+        verifyRouteLineIsUpdatedWithAlternatives()
+
+        // Wait for route progress to come back with an alternative set.
+        val routeProgressCount = CountDownLatch(1)
 
         runOnMainSync {
-            myResourceIdler.increment()
-            assertEquals(3, routeLineApi.getRoutes().size)
-
-            val soonToBePrimaryRoute = routeLineApi.getRoutes()[1]
-            val updatedRoutes = routeLineApi.getRoutes()
-                .filter { it != soonToBePrimaryRoute }
-                .toMutableList()
-                .also {
-                    it.add(0, soonToBePrimaryRoute)
-                }
+            // Create routes where an alternative is set to the primary route position.
+            val updatedRoutes = routeLineApi.getRoutes().toMutableList()
+            val alternative = updatedRoutes[1]
+            updatedRoutes[1] = updatedRoutes[0]
+            updatedRoutes[0] = alternative
+
+            // Update the route line api and MapboxNavigation with an alternative primary.
             val updatedRouteLines = updatedRoutes.map { RouteLine(it, null) }
             routeLineApi.setRoutes(updatedRouteLines) { result ->
                 routeLineView.renderRouteDrawData(
                     activity.mapboxMap.getStyle()!!,
                     result
                 )
             }
-
             mapboxNavigation.setRoutes(updatedRoutes)
-            mapboxNavigation.registerRouteProgressObserver(
-                object : RouteProgressObserver {
-                    override fun onRouteProgressChanged(routeProgress: RouteProgress) {
-                        // only need one route progress for this test
-                        mapboxNavigation.unregisterRouteProgressObserver(this)
-
-                        assertEquals(routeLineApi.getRoutes()[0], routeProgress.route)
-                        routeProgressIdlingResource.unregister()
-                        myResourceIdler.decrement()
-                    }
-                }
-            )
+
+            // Observe route progress and verify the alternative is now the primary route.
+            mapboxNavigation.registerRouteProgressObserver { routeProgress ->
+                assertEquals(alternative, routeProgress.route)
+                assertEquals(alternative, routeLineApi.getRoutes()[0])
```

Comment:
com.mapbox.navigation.instrumentation_tests.ui.routeline.AlternativeRouteSelectionTest$expect_route_apis_to_be_updated_with_alternatives$1.invoke$lambda-2(AlternativeRouteSelectionTest.kt:84)
	at 

it fails this line. RouteProgress is updated. This actually appears to be catching an issue with `routeLineApi.setRoutes`

Reply:
Also suggested by @cafesilencio , moved the assertion into the setRoute callback to avoid the race condition